pkg/controller/oam/applicationconfiguration/render.go

```diff
@@ -0,0 +1,227 @@
+limitations under the License.
+*/
+
+package applicationconfiguration
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplane/crossplane-runtime/pkg/fieldpath"
+
+	"github.com/crossplane/crossplane/apis/oam/v1alpha2"
+)
+
+// Render error strings.
+const (
+	errUnmarshalWorkload = "cannot unmarshal workload"
+	errUnmarshalTrait    = "cannot unmarshal trait"
+)
+
+// Render error format strings.
+const (
+	errFmtGetComponent     = "cannot get component %q"
+	errFmtResolveParams    = "cannot resolve parameter values for component %q"
+	errFmtRenderWorkload   = "cannot render workload for component %q"
+	errFmtRenderTrait      = "cannot render trait for component %q"
+	errFmtSetParam         = "cannot set parameter %q"
+	errFmtUnsupportedParam = "unsupported parameter %q"
+	errFmtRequiredParam    = "required parameter %q not specified"
+)
+
+// A ComponentRenderer renders an ApplicationConfiguration's Components into
+// workloads and traits.
+type ComponentRenderer interface {
+	Render(ctx context.Context, ac *v1alpha2.ApplicationConfiguration) ([]Workload, error)
+}
+
+// A ComponentRenderFn renders an ApplicationConfiguration's Components into
+// workloads and traits.
+type ComponentRenderFn func(ctx context.Context, ac *v1alpha2.ApplicationConfiguration) ([]Workload, error)
+
+// Render an ApplicationConfiguration's Components into workloads and traits.
+func (fn ComponentRenderFn) Render(ctx context.Context, ac *v1alpha2.ApplicationConfiguration) ([]Workload, error) {
+	return fn(ctx, ac)
+}
+
+type components struct {
+	client   client.Reader
+	params   ParameterResolver
+	workload ResourceRenderer
+	trait    ResourceRenderer
+}
+
+func (r *components) Render(ctx context.Context, ac *v1alpha2.ApplicationConfiguration) ([]Workload, error) {
+	workloads := make([]Workload, len(ac.Spec.Components))
+	for i, acc := range ac.Spec.Components {
+
+		c := &v1alpha2.Component{}
+		nn := types.NamespacedName{Namespace: ac.GetNamespace(), Name: acc.ComponentName}
+		if err := r.client.Get(ctx, nn, c); err != nil {
+			return nil, errors.Wrapf(err, errFmtGetComponent, acc.ComponentName)
+		}
+
+		p, err := r.params.Resolve(c.Spec.Parameters, acc.ParameterValues)
+		if err != nil {
+			return nil, errors.Wrapf(err, errFmtResolveParams, acc.ComponentName)
+		}
+
+		w, err := r.workload.Render(c.Spec.Workload.Raw, p...)
+		if err != nil {
+			return nil, errors.Wrapf(err, errFmtRenderWorkload, acc.ComponentName)
+		}
+
+		ref := metav1.NewControllerRef(ac, v1alpha2.ApplicationConfigurationGroupVersionKind)
+		w.SetOwnerReferences([]metav1.OwnerReference{*ref})
+		w.SetNamespace(ac.GetNamespace())
+
+		traits := make([]unstructured.Unstructured, len(acc.Traits))
+		for i, ct := range acc.Traits {
```

Comment:
Another thing I knew but didn't mention is that we didn't check if the trait can apply to this workload here as the information is in the traitDefinition.

Reply:
https://github.com/crossplane/crossplane/pull/1321/files#diff-d7fc5ced21195b2b1a109ef569d402f3R160

Right - my thinking is that validation of whether something is allowed (i.e. whether a trait may apply to a particular kind of workload) should be done using a validating admission control webhook, not a control. This lets us keep validation logic decoupled from reconcilation logic.